internalsdk/ios/ios.go

```diff
@@ -0,0 +1,298 @@
+
+	// Reconfigure forces the ClientWriter to update its configuration
+	Reconfigure()
+
+	Close() error
+}
+
+type StatsTracker interface {
+	UpdateStats(string, string, string, int, int, bool)
+}
+
+type cw struct {
+	ipStack        io.WriteCloser
+	client         *iosClient
+	dialer         *bandit.BanditDialer
+	ipp            ipproxy.Proxy
+	quotaTextPath  string
+	lastSavedQuota time.Time
+}
+
+func (c *cw) Write(b []byte) (int, error) {
+	_, err := c.ipStack.Write(b)
+
+	result := 0
+	return result, err
+}
+
+func (c *cw) Reconfigure() {
+	dialers, err := c.client.loadDialers()
+	if err != nil {
+		// this causes the NetworkExtension process to die. Since the VPN is configured as "on-demand",
+		// the OS will automatically restart the service, at which point we'll read the new config anyway.
+		panic(log.Errorf("Unable to load dialers on reconfigure: %v", err))
+	}
+
+	c.dialer, err = bandit.New(dialers)
+	if err != nil {
+		log.Errorf("Unable to create dialer on reconfigure: %v", err)
+	}
+}
+
+func (c *cw) Close() error {
+	if c.client != nil {
+		c.client.packetsOut.Close()
+	}
+	return nil
+}
+
+type iosClient struct {
+	packetsOut io.WriteCloser
+	udpDialer  UDPDialer
+
+	memChecker      MemChecker
+	configDir       string
+	ipp             ipproxy.Proxy
+	mtu             int
+	capturedDNSHost string
+	realDNSHost     string
+	uc              common.UserConfig
+	tcpHandler      *proxiedTCPHandler
+	udpHandler      *directUDPHandler
+
+	clientWriter    *cw
+	memoryAvailable int64
+	started         time.Time
+	statsTracker    StatsTracker
+}
+
+func Client(packetsOut Writer, udpDialer UDPDialer, memChecker MemChecker, configDir string, mtu int, capturedDNSHost, realDNSHost string, statsTracker StatsTracker) (ClientWriter, error) {
+	log.Debug("Creating new iOS client")
+	if mtu <= 0 {
+		log.Debug("Defaulting MTU to 1500")
+		mtu = 1500
+	}
+
+	c := &iosClient{
+		packetsOut: newWriterAdapter(packetsOut),
+		memChecker: memChecker,
+		configDir:  configDir,
+		//ipp:             ipp,
+		mtu:             mtu,
+		udpDialer:       udpDialer,
+		capturedDNSHost: capturedDNSHost,
+		realDNSHost:     realDNSHost,
+		started:         time.Now(),
+		statsTracker:    statsTracker,
+	}
+	optimizeMemoryUsage(&c.memoryAvailable)
+	go c.gcPeriodically()
+	go c.logMemory()
+
+	return c.start()
+}
+
+func (c *iosClient) start() (ClientWriter, error) {
+	if err := c.loadUserConfig(); err != nil {
+		return nil, log.Errorf("error loading user config: %v", err)
+	}
+	log.Debugf("Running client at config path '%v'", c.configDir)
+	start := time.Now()
+	log.Debugf("User config process start at %v", start)
+	dialers, err := c.loadDialers()
+	if err != nil {
+		return nil, err
+	}
+	if len(dialers) == 0 {
+		return nil, errors.New("No dialers found")
+	}
+	tracker := stats.NewTracker()
+	dialer, err := bandit.NewWithStats(dialers, tracker)
+	if err != nil {
+		return nil, err
+	}
+	go func() {
+		tracker.AddListener(func(st stats.Stats) {
```

Comment:
@atavism Every time we get successfully dali it will call AddListener, so it changes server location again and again.
So if I open some website or do any kind of surfing  it will AddListener so many time, It just add lot of unnecessary calls to Tunnel->App->Go ,

Is there any way we can optimise this?

Reply:
The `tracker.AddListener` code here should only be called once, when we start the iOS client in the PacketTunnelProvider. What makes you think it is being called multiple times? Are you likewise seeing [startFlashlight](https://github.com/getlantern/lantern-client/blob/atavism/ios/ios/Tunnel/PacketTunnelProvider.swift#L115) called multiple times?